### PR TITLE
Make sure `inspect.iscoroutinefunction` works on coroutines decorated with `@validate_call`

### LINKED ISSUE #10370: Support async functions with `@validate_call` decorator
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I'm writing a project using `asgiref`.

When I use `async_to_sync`, I got a error ` UserWarning: async_to_sync was passed a non-async-marked callable`.

I think it's due to `@validate_call` decorator because it's not appear if i delete `@validate_call`.

For example, here is a simple code. The correct output is `True` but actual output is `False`.  How can we fix this bug?

### Example Code

```Python
import asyncio
import inspect

from pydantic import validate_call


@validate_call(validate_return=True)
async def delay_number(n: int) -> int:
    await asyncio.sleep(1)
    return n


if __name__ == "__main__":
    print(inspect.iscoroutinefunction(delay_number))
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.9.1
        pydantic-core version: 2.23.3
          pydantic-core build: profile=release pgo=false
                 install path: C:\Users\movis\miniconda3\envs\volans_env\Lib\site-packages\pydantic
               python version: 3.11.9 | packaged by Anaconda, Inc. | (main, Apr 19 2024, 16:40:41) [MSC v.1916 64 bit (AMD64)]
                     platform: Windows-10-10.0.22631-SP0
             related packages: fastapi-0.112.4 typing_extensions-4.12.2
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by MovisLi**
## Change Summary

Add judgement for validate_call decorator. Now we can get `True` return from `inpect.iscoroutinefunction()`.

## Related issue number

fix #10370 

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by MovisLi** [comment]
@sydney-runkle , please review.

File: pydantic/validate_call_decorator.py [hunk 1]
```diff
@@ -59,9 +60,17 @@ def validate(function: AnyCallableT) -> AnyCallableT:
...
+            return await validate_call_wrapper(*args, **kwargs)
+
         wrapper_function.raw_function = function  # type: ignore
+        async_wrapper_function.raw_function = function  # type: ignore
 
-        return wrapper_function  # type: ignore
+        if inspect.iscoroutinefunction(function):
+            return async_wrapper_function  # type: ignore
+        else:
+            return wrapper_function  # type: ignore
```
Comment (sydney-runkle) [inline pydantic/validate_call_decorator.py:73]: Let's add the function definitions (`wrapper_function` and `async_wrapper_function`) to the relevant parts of this conditional. Can also move the `.raw_function` assignments there as well.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looking good, could you please add a test as well? Thanks!

File: pydantic/validate_call_decorator.py (hunk 1 above)
Comment (MovisLi) [reply pydantic/validate_call_decorator.py:73]: ok

File: tests/test_validate_call.py
Comment (Viicos) [inline tests/test_validate_call.py:1]: Could you please revert the formatting changes, if not necessary?

**Comment by Viicos** [comment]
> I don't quite understand what changes need to be made to the test cases. I add some codes to check if it is a coroutine the async function under decorator. Please check it, thank you!

Something like:

```python
import inspect

@validate_call
async def test():
    pass

assert inspect.iscoroutinefunction(test)
```

File: tests/test_validate_call.py
Comment (MovisLi) [reply tests/test_validate_call.py:1]: Sorry, I don't  know why ruff format this codes. But I'll try to revert it.

**Comment by sydney-runkle** [comment]
Happy to do another review here after we:

* revert formatting changes
* add a test like @Viicos sugested
* Move the function definitions into the conditional blocks, see below:

currently:

```
# function definitions

if condition:
    return func1
else:
    return func2
```

desired:

```
if condition:
    # define func1
    return func1
else:
    # define func2
    return func2
```

Thanks, great work here!

**Comment by MovisLi** [comment]
@Viicos @sydney-runkle I finish all @sydney-runkle mentioned. Please check it, thank you!

**Review by MovisLi — COMMENTED** [review]
check

**Review by sydney-runkle — APPROVED** [review]
Awesome job, thanks for the help here!!

**Pull request merged**